File: 398. Random Pick Index/solution.java
class Solution {
    private Map<Integer, List<Integer>> data;
    private Random random;
    public Solution(int[] nums) {
        data = new HashMap<>();
        for (int i = 0;i < nums.length; i++) {
            List<Integer> indexes = data.get(nums[i]);
            if (indexes != null) {
                indexes.add(i);
            } else {
                indexes = new ArrayList<>();
                indexes.add(i);
                data.put(nums[i], indexes);
            }
        }
        // data.entrySet().stream().forEach(entry -> System.out.println("key: " + entry.getKey() + " value: " + entry.getValue()));
        random = new Random();
    }
    
    public int pick(int target) {
        List<Integer> indexes = data.get(target);
        return indexes.get(random.nextInt(indexes.size()));
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int param_1 = obj.pick(target);
 */

File: 398. Random Pick Index/solution.py
class Solution:

    def __init__(self, nums: List[int]):
        self.data = {}
        for idx in range(len(nums)):
            if nums[idx] in self.data:
                self.data[nums[idx]].append(idx)
            else:
                self.data[nums[idx]] = [idx]
        print(self.data)

    def pick(self, target: int) -> int:
        # print(target)
        return self.data[target][random.randint(0, len(self.data[target]) - 1)]


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.pick(target)
